Stub unavailable source when reporting Starlark errors

pivotaljohn opened the pull request:
Addresses:
- #156
- #687

jtigger committed:
Stub unavailable source when reporting Starlark errors

cppforlife commented on pkg/template/compiled_template_error.go:
```diff
@@ -156,32 +174,39 @@ func (e CompiledTemplateMultiError) buildPos(pos syntax.Position) CompiledTempla
 
-	ct, err := e.loader.FindCompiledTemplate(pos.Filename())
+	var sc sourceCode
+	sc, err := e.loader.FindCompiledTemplate(pos.Filename())
```
should we switch away from err to bool (found or not found)?

pivotaljohn replied:
👍🏻  Avoid using `error`s for flow control. With this PR, it's no longer an error.

pivotaljohn replied:
I decided that `nil` is sufficient in this case (rather than two return values). `nil` is never a valid "found" value.

cppforlife commented on pkg/template/compiled_template_error.go:
```diff
@@ -148,6 +148,24 @@ func (e CompiledTemplateMultiError) buildEvalErr(err *starlark.EvalError) Compil
+
+func (s unavailableSource) CodeAtLine(pos *filepos.Position) *Line {
+	return &Line{
+		Instruction: Instruction{},
```
nit: not necessary since its zero value and not a pointer.

cppforlife commented on pkg/template/compiled_template_error.go:
```diff
@@ -148,6 +148,24 @@ func (e CompiledTemplateMultiError) buildEvalErr(err *starlark.EvalError) Compil
+		SourceLine: &SourceLine{
+			Position:  pos,
+			Content:   "(unavailable)",
+			Selection: nil,
```
same as above

cppforlife commented on pkg/template/compiled_template_error.go:
```diff
@@ -156,32 +174,39 @@ func (e CompiledTemplateMultiError) buildPos(pos syntax.Position) CompiledTempla
 	}
 
-	line := ct.CodeAtLine(filepos.NewPosition(int(pos.Line)))
+	p := filepos.NewPosition(int(pos.Line))
```
should we add a new constructor that add file and line?

pivotaljohn replied:
👍🏻  In fact, long overdue.

jtigger committed:
Construct a Position with line number and filename

jtigger committed:
When template is not found, return nil instead of an error

- record the fact that we intend for this to be a temporary fallback
  strategy: that we still aspire to supply offending lines of source
  even if not part of the Library under evaluation.

pivotaljohn commented on pkg/template/compiled_template_error.go:
```diff
@@ -175,11 +178,9 @@ func (e CompiledTemplateMultiError) buildPos(pos syntax.Position) CompiledTempla
-		// - a function from another Library
-		// - Starlark sourced from within ytt (e.g. out-of-the-box validation functions)
+	sc = e.loader.FindCompiledTemplate(pos.Filename())
+	if reflect.ValueOf(sc).IsNil() {
```
required idiom as `sc` is an interface, not a pointer.
